dsh_sdk/src/mqtt_token_fetcher.rs

```diff
@@ -0,0 +1,359 @@
+    }
+
+    fn is_valid(&self) -> bool {
+        let current_unixtime = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("SystemTime before UNIX EPOCH!")
+            .as_secs() as i32;
+        self.exp >= current_unixtime - 5
+    }
+}
+
+/// Represents attributes associated with a Rest token.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "kebab-case")]
+struct RestTokenAttributes {
+    gen: i64,
+    pub endpoint: String,
+    iss: String,
+    pub claims: RestClaims,
+    pub exp: i64,
+    tenant_id: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct RestClaims {
+    #[serde(rename = "datastreams/v0/mqtt/token")]
+    datastreams_token: DatastreamsData,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct DatastreamsData {}
+
+/// Represents a rest token with its raw value and attributes.
+#[derive(Serialize, Deserialize, Debug)]
+struct RestToken {
+    raw_token: String, //impl setter/getter
+    exp: i64,
+}
+
+impl RestToken {
+    /// Creates a new instance of `RestToken` from a raw token string.
+    ///
+    /// # Arguments
+    ///
+    /// * `raw_token` - The raw token string.
+    ///
+    /// # Returns
+    ///
+    /// A Result containing the created RestToken or an error.
+    async fn new(tenant: &String, api_key: &String, env: &Platform) -> Result<RestToken, DshError> {
+        let raw_token = Self::create_rest_token(tenant, api_key, env).await.unwrap();
+
+        let header_payload = extract_header_and_payload(&raw_token)?;
+
+        let decoded_token = decode_payload(header_payload)?;
+
+        let token_attributes = RestToken::parse_token_attributes(&decoded_token)?;
+        let token = RestToken {
+            raw_token,
+            exp: token_attributes.exp,
+        };
+
+        Ok(token)
+    }
+
+    fn is_valid(&self) -> bool {
+        let current_unixtime = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("SystemTime before UNIX EPOCH!")
+            .as_secs() as i64;
+        self.exp >= current_unixtime - 5
+    }
+
+    fn parse_token_attributes(decoded_token: &[u8]) -> Result<RestTokenAttributes, DshError> {
+        let res = serde_json::from_slice(decoded_token).map_err(DshError::JsonError);
+        res
+    }
+
+    async fn create_rest_token(
+        tenant: &String,
+        api_key: &String,
+        env: &Platform,
+    ) -> Result<String, DshError> {
+        let json_body = json!({"tenant": tenant});
+
+        const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+
+        let reqwest_client = Client::builder()
+            .timeout(DEFAULT_TIMEOUT)
+            .http1_only()
+            .build()
+            .expect("Failed to build reqwest client");
+
+        let rest_client = reqwest_client;
+        let response = rest_client
+            .post(env.endpoint_rest_token())
+            .header("apikey", &api_key.to_string())
```

Comment:
unnecessary borrow and clone
```suggestion
            .header("apikey", api_key)
```